include/nbl/video/IDescriptorPool.h

```diff
@@ -29,12 +33,156 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
+            m_storageImageStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE] + m_maxDescriptorCount[asset::EDT_INPUT_ATTACHMENT]);
+            m_UBO_SSBOStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_UTB_STBStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_accelerationStructureStorage = std::make_unique<StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
+        }
+
+        ~IDescriptorPool()
+        {
+            for (auto i = 0; i < asset::EDT_COUNT; ++i)
+            {
+                if ((m_flags & IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT) && (m_maxDescriptorCount[i] > 0))
+                    _NBL_ALIGNED_FREE(m_generalAllocatorReservedSpace[i]);
+            }
+        }
+
+        // This will return the offset into the pool's descriptor storage. These offsets will be combined
+        // later with base memory addresses to get the actual memory adress where we put the core::smart_refctd_ptr<const IDescriptor>.
+        uint32_t allocateDescriptors(const asset::E_DESCRIPTOR_TYPE type, const uint32_t count)
+        {
+            const uint32_t bytesToAllocate = count * sizeof(void*);
+
+            uint32_t offset;
+            uint32_t invalidAddress;
+            if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+            {
+                offset = m_generalAllocators[type].alloc_addr(bytesToAllocate, 1u);
+                invalidAddress = core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address;
+            }
+            else
+            {
+                offset = m_linearAllocators[type].alloc_addr(bytesToAllocate, 1u);
+                invalidAddress = core::LinearAddressAllocator<uint32_t>::invalid_address;
+            }
+
+            return (offset == invalidAddress) ? ~0u : offset;
+        }
+
+        void freeDescriptors(const uint32_t count, void* descriptors, const asset::E_DESCRIPTOR_TYPE type)
+        {
+            assert(m_flags & ECF_FREE_DESCRIPTOR_SET_BIT);
+            uint32_t allocatedAddr = static_cast<uint32_t>(reinterpret_cast<uint8_t*>(descriptors) - getDescriptorMemoryBaseAddress(type));
+            m_generalAllocators[type].free_addr(allocatedAddr, count*sizeof(core::smart_refctd_ptr<const asset::IDescriptor>));
+        }
+
+        inline uint8_t* getDescriptorMemoryBaseAddress(const asset::E_DESCRIPTOR_TYPE type) const
```

Comment:
for [Dependency Injection](https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#issuecomment-1263217433) to work, this method needs to be a protected method of `IDescriptorSet`

Yes this means that `IGPUDescriptorSet` requires access to private members of the `IDescriptorPool` so declare it a friend.

Also can operate in `core::smart_refctd_ptr<IDescriptor>` instead of raw byte pointers.

Reply:
yes resolved, DI was a mistake. whats there right now is fine